File: sdk-go/account.go
package boxkite

import (
	"context"
	"encoding/json"
	"net/url"
	"strconv"
)

// Account is the identity for the API key in use (GET /v1/account).
type Account struct {
	ID        string `json:"id"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
}

// Usage is the current usage against fair-use limits (GET /v1/usage), also
// returned inline on CreateSandbox's response.
type Usage struct {
	MonthlySandboxHoursUsed  float64 `json:"monthly_sandbox_hours_used"`
	MonthlySandboxHoursLimit float64 `json:"monthly_sandbox_hours_limit"`
	ConcurrentSandboxes      int     `json:"concurrent_sandboxes"`
	ConcurrentSandboxesLimit int     `json:"concurrent_sandboxes_limit"`
}

// MessageResponse is a generic ack body (e.g. password-reset request,
// which always returns the same message regardless of whether the email
// is registered).
type MessageResponse struct {
	Message string `json:"message"`
}

// Account returns the account identity for the API key in use.
func (c *Client) Account(ctx context.Context) (*Account, error) {
	var out Account
	if err := c.doJSON(ctx, "GET", "/v1/account", nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// Usage returns current usage against fair-use limits.
func (c *Client) Usage(ctx context.Context) (*Usage, error) {
	var out Usage
	if err := c.doJSON(ctx, "GET", "/v1/usage", nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// RequestPasswordReset requests a password-reset email
// (POST /v1/auth/password-reset/request). Opt-in on the control-plane
// (BOXKITE_PASSWORD_RESET_ENABLED); returns an *APIError with
// StatusCode 404 and Code "feature_disabled" if the deployment hasn't
// enabled it. Always returns the same message whether or not the email is
// registered, so this call can never be used to enumerate accounts.
func (c *Client) RequestPasswordReset(ctx context.Context, email string) (*MessageResponse, error) {
	var out MessageResponse
	body := map[string]string{"email": email}
	if err := c.doJSON(ctx, "POST", "/v1/auth/password-reset/request", body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ConfirmPasswordReset consumes a single-use token minted by
// RequestPasswordReset and sets a new password
// (POST /v1/auth/password-reset/confirm). Also revokes every outstanding
// refresh token for the account, if refresh tokens are enabled
// server-side.
func (c *Client) ConfirmPasswordReset(ctx context.Context, token, newPassword string) (*MessageResponse, error) {
	var out MessageResponse
	body := map[string]string{"token": token, "new_password": newPassword}
	if err := c.doJSON(ctx, "POST", "/v1/auth/password-reset/confirm", body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// VerifyEmail consumes a single-use email-verification token
// (POST /v1/auth/verify-email), opt-in
// (BOXKITE_EMAIL_VERIFICATION_ENABLED).
func (c *Client) VerifyEmail(ctx context.Context, token string) (*MessageResponse, error) {
	var out MessageResponse
	body := map[string]string{"token": token}
	if err := c.doJSON(ctx, "POST", "/v1/auth/verify-email", body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ResendVerification re-sends the verification email for the
// dashboard-JWT-authenticated account (POST /v1/auth/resend-verification).
// accessToken is a dashboard session token (the JWT returned by
// /v1/auth/login or /v1/auth/signup) -- a different, non-interchangeable
// credential type from this Client's own apiKey, so it overrides this
// call's Authorization header rather than using the Client's apiKey.
func (c *Client) ResendVerification(ctx context.Context, accessToken string) (*MessageResponse, error) {
	var out MessageResponse
	opts := &requestOptions{authOverride: accessToken}
	if err := c.doJSON(ctx, "POST", "/v1/auth/resend-verification", nil, &out, opts); err != nil {
		return nil, err
	}
	return &out, nil
}

// TokenPair is the response shape from RefreshToken -- a brand new
// access_token + refresh_token pair, plus the account identity.
type TokenPair struct {
	AccessToken  string  `json:"access_token"`
	TokenType    string  `json:"token_type"`
	ExpiresIn    int     `json:"expires_in"`
	RefreshToken *string `json:"refresh_token"`
	Account      Account `json:"account"`
}

// RefreshToken exchanges a still-valid refresh token for a brand new
// access_token + refresh_token pair (POST /v1/auth/refresh), opt-in
// (BOXKITE_REFRESH_TOKENS_ENABLED). Revokes the presented token in the
// same request (rotation, not reuse) -- store the new RefreshToken from
// the response and discard the one presented here.
func (c *Client) RefreshToken(ctx context.Context, refreshToken string) (*TokenPair, error) {
	var out TokenPair
	body := map[string]string{"refresh_token": refreshToken}
	if err := c.doJSON(ctx, "POST", "/v1/auth/refresh", body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// Logout revokes one refresh token immediately (POST /v1/auth/logout),
// opt-in (BOXKITE_REFRESH_TOKENS_ENABLED). Always succeeds (204) whether
// or not the token was valid -- never leaks which.
func (c *Client) Logout(ctx context.Context, refreshToken string) error {
	body := map[string]string{"refresh_token": refreshToken}
	return c.doJSON(ctx, "POST", "/v1/auth/logout", body, nil, nil)
}

// AllowedCommandRule is one account-level command allowlist rule -- either
// a bare command name (Command set, ArgsAllow/ArgsDeny both empty) or a
// command name plus argument allow/deny regex lists. The control-plane's
// own schema (AllowedCommandsRequest.rules: list[str | AllowedCommandRule]
// in control-plane/src/control_plane/schemas.py) accepts and returns rules
// in either shape -- UnmarshalJSON below decodes both into the same Go
// struct; MarshalJSON always emits the object form, which the
// control-plane accepts identically to the bare-string form for a rule
// with no argument constraints.
type AllowedCommandRule struct {
	Command   string   `json:"command"`
	ArgsAllow []string `json:"args_allow,omitempty"`
	ArgsDeny  []string `json:"args_deny,omitempty"`
}

// allowedCommandRuleAlias avoids infinite recursion when UnmarshalJSON
// below delegates the object-shaped case back to encoding/json.
type allowedCommandRuleAlias AllowedCommandRule

// UnmarshalJSON accepts either a bare JSON string (a plain command name,
// no argument constraints) or a JSON object
// ({"command", "args_allow"?, "args_deny"?}).
func (r *AllowedCommandRule) UnmarshalJSON(data []byte) error {
	var command string
	if err := json.Unmarshal(data, &command); err == nil {
		r.Command = command
		r.ArgsAllow = nil
		r.ArgsDeny = nil
		return nil
	}
	var alias allowedCommandRuleAlias
	if err := json.Unmarshal(data, &alias); err != nil {
		return err
	}
	*r = AllowedCommandRule(alias)
	return nil
}

// AllowedCommandsResponse is the body shape shared by
// GetAllowedCommands/SetAllowedCommands.
type AllowedCommandsResponse struct {
	Rules []AllowedCommandRule `json:"rules"`
}

// GetAllowedCommands returns the current per-account command allowlist
// (GET /v1/account/allowed-commands). An empty Rules means unrestricted --
// the default for every account. This allowlist is an opt-in guardrail,
// not a sandbox-escape boundary.
func (c *Client) GetAllowedCommands(ctx context.Context) (*AllowedCommandsResponse, error) {
	var out AllowedCommandsResponse
	if err := c.doJSON(ctx, "GET", "/v1/account/allowed-commands", nil, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// SetAllowedCommands replaces the per-account command allowlist wholesale
// (PUT /v1/account/allowed-commands). rules must be non-empty -- use
// ClearAllowedCommands to reset to unrestricted.
func (c *Client) SetAllowedCommands(ctx context.Context, rules []AllowedCommandRule) (*AllowedCommandsResponse, error) {
	var out AllowedCommandsResponse
	body := AllowedCommandsResponse{Rules: rules}
	if err := c.doJSON(ctx, "PUT", "/v1/account/allowed-commands", body, &out, nil); err != nil {
		return nil, err
	}
	return &out, nil
}

// ClearAllowedCommands removes the per-account command allowlist
// (DELETE /v1/account/allowed-commands), back to the unrestricted default.
func (c *Client) ClearAllowedCommands(ctx context.Context) error {
	return c.doJSON(ctx, "DELETE", "/v1/account/allowed-commands", nil, nil, nil)
}

func boolQueryValue(v bool) string {
	return strconv.FormatBool(v)
}

func newQuery() url.Values {
	return url.Values{}
}

File: sdk-rust/src/account.rs
//! Account/usage introspection, dashboard auth-flow helpers, and the
//! per-account command allowlist. Mirrors `sdk-python`'s `account`/`usage`/
//! `request_password_reset`/`confirm_password_reset`/`verify_email`/
//! `resend_verification`/`refresh_token`/`logout`/`get_allowed_commands`/
//! `set_allowed_commands`/`clear_allowed_commands` and `sdk-go`'s
//! same-named methods.

use reqwest::Method;
use serde::de::{self, Deserializer};
use serde::{Deserialize, Serialize};

use crate::client::Client;
use crate::error::BoxkiteError;
use crate::sandboxes::UsageSummary;

/// The account identity for the API key in use (`GET /v1/account`).
#[derive(Debug, Clone, Deserialize)]
pub struct Account {
    pub id: String,
    pub email: String,
    pub created_at: String,
}

/// A generic ack body (e.g. a password-reset request, which returns the same
/// message whether or not the email is registered).
#[derive(Debug, Clone, Deserialize)]
pub struct MessageResponse {
    pub message: String,
}

/// A fresh `access_token` + `refresh_token` pair plus the account identity,
/// returned by [`Client::refresh_token`].
#[derive(Debug, Clone, Deserialize)]
pub struct TokenPair {
    pub access_token: String,
    pub token_type: String,
    pub expires_in: i64,
    pub refresh_token: Option<String>,
    pub account: Account,
}

/// One account-level command-allowlist rule -- either a bare command name
/// (`args_allow`/`args_deny` both empty) or a command name plus argument
/// allow/deny regex lists.
///
/// The control-plane accepts and returns rules in either shape (a bare JSON
/// string or a `{command, args_allow?, args_deny?}` object); this type's
/// `Deserialize` decodes both, while its `Serialize` always emits the object
/// form (which the control-plane accepts identically). Mirrors `sdk-go`'s
/// `AllowedCommandRule` custom marshaling.
#[derive(Debug, Clone, Serialize)]
pub struct AllowedCommandRule {
    pub command: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args_allow: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args_deny: Vec<String>,
}

impl AllowedCommandRule {
    /// A bare rule: just a command name, no argument constraints.
    pub fn new(command: impl Into<String>) -> Self {
        Self {
            command: command.into(),
            args_allow: Vec::new(),
            args_deny: Vec::new(),
        }
    }

    pub fn args_allow<I, S>(mut self, patterns: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.args_allow = patterns.into_iter().map(Into::into).collect();
        self
    }

    pub fn args_deny<I, S>(mut self, patterns: I) -> Self
    where
        I: IntoIterator<Item = S>,
        S: Into<String>,
    {
        self.args_deny = patterns.into_iter().map(Into::into).collect();
        self
    }
}

impl<'de> Deserialize<'de> for AllowedCommandRule {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(untagged)]
        enum Raw {
            Bare(String),
            Full {
                command: String,
                #[serde(default)]
                args_allow: Vec<String>,
                #[serde(default)]
                args_deny: Vec<String>,
            },
        }
        match Raw::deserialize(deserializer).map_err(de::Error::custom)? {
            Raw::Bare(command) => Ok(Self::new(command)),
            Raw::Full {
                command,
                args_allow,
                args_deny,
            } => Ok(Self {
                command,
                args_allow,
                args_deny,
            }),
        }
    }
}

/// Body shape shared by [`Client::get_allowed_commands`]/
/// [`Client::set_allowed_commands`]. An empty `rules` means unrestricted --
/// the default for every account.
#[derive(Debug, Clone, Default, Deserialize, Serialize)]
pub struct AllowedCommandsResponse {
    #[serde(default)]
    pub rules: Vec<AllowedCommandRule>,
}

impl Client {
    /// `GET /v1/account` -- the account identity for the API key in use.
    pub async fn account(&self) -> Result<Account, BoxkiteError> {
        let builder = self.request(Method::GET, "/v1/account");
        self.send(builder).await
    }

    /// `GET /v1/usage` -- current usage against this account's fair-use
    /// limits (the same shape returned inline on [`Client::create_sandbox`]).
    pub async fn usage(&self) -> Result<UsageSummary, BoxkiteError> {
        let builder = self.request(Method::GET, "/v1/usage");
        self.send(builder).await
    }

    /// `POST /v1/auth/password-reset/request` -- request a password-reset
    /// email. Opt-in server-side (`BOXKITE_PASSWORD_RESET_ENABLED`); 404s
    /// with code `feature_disabled` if the deployment hasn't enabled it.
    /// Always returns the same message whether or not the email is
    /// registered, so it can't be used to enumerate accounts.
    pub async fn request_password_reset(
        &self,
        email: &str,
    ) -> Result<MessageResponse, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            email: &'a str,
        }
        let builder = self
            .request(Method::POST, "/v1/auth/password-reset/request")
            .json(&Body { email });
        self.send(builder).await
    }

    /// `POST /v1/auth/password-reset/confirm` -- consume a single-use token
    /// from [`Client::request_password_reset`] and set a new password. Also
    /// revokes every outstanding refresh token for the account, if refresh
    /// tokens are enabled server-side.
    pub async fn confirm_password_reset(
        &self,
        token: &str,
        new_password: &str,
    ) -> Result<MessageResponse, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            token: &'a str,
            new_password: &'a str,
        }
        let builder = self
            .request(Method::POST, "/v1/auth/password-reset/confirm")
            .json(&Body {
                token,
                new_password,
            });
        self.send(builder).await
    }

    /// `POST /v1/auth/verify-email` -- consume a single-use email-verification
    /// token. Opt-in server-side (`BOXKITE_EMAIL_VERIFICATION_ENABLED`).
    pub async fn verify_email(&self, token: &str) -> Result<MessageResponse, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            token: &'a str,
        }
        let builder = self
            .request(Method::POST, "/v1/auth/verify-email")
            .json(&Body { token });
        self.send(builder).await
    }

    /// `POST /v1/auth/resend-verification` -- re-send the verification email
    /// for the dashboard-JWT-authenticated account. `access_token` is a
    /// dashboard session token (the JWT returned by `/v1/auth/login` or
    /// `/v1/auth/signup`) -- a different, non-interchangeable credential type
    /// from this client's own `api_key`, so it overrides this call's
    /// `Authorization` header rather than using the client's key.
    pub async fn resend_verification(
        &self,
        access_token: &str,
    ) -> Result<MessageResponse, BoxkiteError> {
        let builder =
            self.request_with_auth(Method::POST, "/v1/auth/resend-verification", access_token);
        self.send(builder).await
    }

    /// `POST /v1/auth/refresh` -- exchange a still-valid refresh token for a
    /// brand new `access_token` + `refresh_token` pair. Opt-in server-side
    /// (`BOXKITE_REFRESH_TOKENS_ENABLED`). Revokes the presented token in the
    /// same request (rotation, not reuse) -- store the new `refresh_token`
    /// from the response and discard the one presented here.
    pub async fn refresh_token(&self, refresh_token: &str) -> Result<TokenPair, BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            refresh_token: &'a str,
        }
        let builder = self
            .request(Method::POST, "/v1/auth/refresh")
            .json(&Body { refresh_token });
        self.send(builder).await
    }

    /// `POST /v1/auth/logout` -- revoke one refresh token immediately. Opt-in
    /// server-side (`BOXKITE_REFRESH_TOKENS_ENABLED`). Always succeeds (204)
    /// whether or not the token was valid -- never leaks which.
    pub async fn logout(&self, refresh_token: &str) -> Result<(), BoxkiteError> {
        #[derive(Serialize)]
        struct Body<'a> {
            refresh_token: &'a str,
        }
        let builder = self
            .request(Method::POST, "/v1/auth/logout")
            .json(&Body { refresh_token });
        self.send_no_content(builder).await
    }

    /// `GET /v1/account/allowed-commands` -- the current per-account command
    /// allowlist. An empty `rules` means unrestricted (the default). This
    /// allowlist is an opt-in guardrail, not a sandbox-escape boundary.
    pub async fn get_allowed_commands(&self) -> Result<AllowedCommandsResponse, BoxkiteError> {
        let builder = self.request(Method::GET, "/v1/account/allowed-commands");
        self.send(builder).await
    }

    /// `PUT /v1/account/allowed-commands` -- replace the per-account command
    /// allowlist wholesale. `rules` must be non-empty; use
    /// [`Client::clear_allowed_commands`] to reset to unrestricted.
    pub async fn set_allowed_commands(
        &self,
        rules: Vec<AllowedCommandRule>,
    ) -> Result<AllowedCommandsResponse, BoxkiteError> {
        let body = AllowedCommandsResponse { rules };
        let builder = self
            .request(Method::PUT, "/v1/account/allowed-commands")
            .json(&body);
        self.send(builder).await
    }

    /// `DELETE /v1/account/allowed-commands` -- remove the per-account
    /// command allowlist, back to the unrestricted default.
    pub async fn clear_allowed_commands(&self) -> Result<(), BoxkiteError> {
        let builder = self.request(Method::DELETE, "/v1/account/allowed-commands");
        self.send_no_content(builder).await
    }
}
